Implement minimal textmate language definition

junlarsen opened the pull request:
TextMate is the go-to definition language for syntax highlighting for editors lack dedicated language support for a given programming language. This patch contains a very minimal, yet functional TextMate bundle definition for Carbon.

This allows Carbon users to import the TextMate bundle into their editors to get basic syntax highlighting for Carbon code until we have a more solid specification for the language (that's when the fun with dedicated plugins/extensions begins!).

**Support**
- String literals with escape codes (currently using C highlighting rules)
- Numeric literals in decimal, hexadecimal and binary (lacks _ separator support though)
- Single and multi-line comments
- Highlighting of all? (let me know if some are missing) keywords in the language

Here is an image of the highlighting in action in IntelliJ Dracula mode
![image](https://user-images.githubusercontent.com/42585241/180587912-23b5ae3e-2bdb-49a1-83e8-24e6e0b7cfb5.png)

junlarsen committed:
feat: implement minimal textmate language definition

junlarsen commented on utils/textmate/info.plist:
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
```
I had some troubles with the copyright checker with this one. It suggests I use `#` for comments, but those are not valid XML comments

jonmeow replied:
In /.pre-commit-config.yaml, should be something like (added under `check-copyright`):

```
      - --custom_format
      - '\.js$'
      - '<!--'
      - ''
      - '-->'
```

(note, may not be worth this right now per other comment)

junlarsen commented on utils/textmate/Syntaxes/Carbon.plist:
```diff
@@ -0,0 +1,181 @@
+                <key>match</key>
+                <string>\b(bool|Carbon\.Int|Carbon\.UInt|f16|f32|f64|f128|i8|i16|i32|i64|i128|i256|Slice|String|StringView|Type|u8|u16|u32|u64|u128|u256)\b</string>
+            </dict>
+            <!-- Rule: numbers - derived from c.tmbundle -->
```
The TextMate code is licensed under GPL 3.0, I'm not sure if that interferes with Apache 2.0-with-LLVM-Exception or not?

jonmeow replied:
Please let us get back to you on this. @chandlerc

junlarsen replied:
Great, if there are any issues I'll happily spend some time to make a new one

jonmeow replied:
Yes, this is an issue. For reference, please don't derive code from sources that don't match our license (Apache 2.0 with LLVM-exception -- this is basically us and llvm-project, so fairly limited group).

If you'd like to continue, remove this file and start from scratch. You can read their documentation, but don't copy code.

Here's an independently development highlight syntax that you should use as a basis:
https://github.com/carbon-language/carbon-lang/blob/trunk/utils/highlightjs/highlightjs_carbon_lang.js#L139

(highlightjs may follow a pretty similar structure overall, making it easy to reuse in a lot of this)

junlarsen committed:
doc: write down installation guides for common editors

jonmeow reviewed (commented):
Please DM me on Discord if you need more information on my comment.

junlarsen committed:
fix: add pattern for .plist and remove gpl-3.0 textmate code

junlarsen committed:
feat: introduce new string/number rules based on highlightjs code

junlarsen committed:
fix: allow lowercase a-f characters in hex contexts

junlarsen commented:
Thanks for the feedback! I've replaced the parts that were derived from TextMate's code and used the regular expressions from our highlightjs file instead.

I've also added the extra space to the pre-commit (as we discussed on Discord) (for others; see https://github.com/google/pre-commit-tool-hooks/issues/17)

Let me know if there is anything else

jonmeow commented on utils/textmate/README.md:
```diff
@@ -0,0 +1,31 @@
+[here](https://macromates.com/manual/en/bundles#getting_more_bundles). Clone the
+repository with Git and symlink/copy the `utils/textmate` directory to any of
+the paths TextMate will search through (you can find these paths in the TextMate
+documentation above)
```
```suggestion
documentation above).
```

Punctuation nit

jonmeow commented on utils/textmate/README.md:
```diff
@@ -0,0 +1,31 @@
+the paths TextMate will search through (you can find these paths in the TextMate
+documentation above)
+
+If you are using IntelliJ or a JetBrains-family product, you can find
+documentation on how to install the bundle
+[here](https://www.jetbrains.com/help/idea/textmate.html#import-textmate-bundles).
```
```suggestion
If you are using IntelliJ or a JetBrains-family product,
[there is documentation on how to install the bundle](https://www.jetbrains.com/help/idea/textmate.html#import-textmate-bundles).
```

This is a doc style thing, just avoiding "here" as link text.
https://developers.google.com/style/link-text

junlarsen replied:
Ah that makes a lot of sense, thanks for the link!

jonmeow commented on utils/textmate/README.md:
```diff
@@ -0,0 +1,31 @@
+IntelliJ TextMate Bundle window.
+
+If you are using Atom, you can convert the bundle to an Atom-compatible one. See
+the Atom documentation
+[here](https://flight-manual.atom.io/hacking-atom/sections/converting-from-textmate/)
+on how to do that.
```
```suggestion
[the Atom documentation on how to do that](https://flight-manual.atom.io/hacking-atom/sections/converting-from-textmate/).
```

jonmeow commented on utils/highlightjs/highlightjs_carbon_lang.js:
```diff
@@ -140,7 +140,7 @@ export default function (hljs) {
     variants: [
       { match: /[1-9][_0-9]*(\.[_0-9]+(e[-+]?[1-9][0-9]*)?)?/ },
-      { match: /0x[_0-9A-F]+(\.[_0-9A-F]+(p[-+]?[1-9][0-9]*)?)?/ },
+      { match: /0x[_0-9a-fA-F]+(\.[_0-9a-fA-F]+(p[-+]?[1-9][0-9]*)?)?/ },
```
The regex was correct; we're looking at requiring uppercase for consistency/readability. References:

https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/numeric_literals.md#integer-literals

https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/string_literals.md#escape-sequences

junlarsen replied:
Ok cool, thanks for clarifying - will revert!

jonmeow commented on utils/textmate/README.md:
```diff
@@ -0,0 +1,31 @@
+-->
+
+This directory contains a [TextMate](https://macromates.com/) bundle which can
+be used in various editors such as TextMate, Atom or the IntelliJ family for
```
```suggestion
be used in various editors such as TextMate, Atom, or the IntelliJ family for
```

We like our Oxford commas.

jonmeow reviewed (commented):
Thanks! I think the remaining bits are fairly small.

Note I'm not really testing the textmate config -- I don't know an easy way to do that (I mostly use vscode these days)

junlarsen committed:
fix: revert lowercase hexadecimal digits

junlarsen committed:
fix: address review comments in prose

junlarsen committed:
fix: use uppercase hex in sample

junlarsen commented:
> Thanks! I think the remaining bits are fairly small.
> 
> Note I'm not really testing the textmate config -- I don't know an easy way to do that (I mostly use vscode these days)

Thanks, I've fixed the nits and reverted the a-f hexadecimal thing. Here's a view from IntelliJ with the default theme
![image](https://user-images.githubusercontent.com/42585241/181575778-7a1abea3-bcd4-4504-a5c2-c4307726b833.png)

I tried to look for ways to test, but it seems the "official" bundles don't any means of testing beyond sample files (that they presumably open in their editor and manually test?)

jonmeow commented on utils/textmate/Syntaxes/Carbon.plist:
```diff
@@ -0,0 +1,173 @@
+            <dict>
+                <key>patterns</key>
+                <array>
+                    <!-- Multi-line comments -->
+                    <dict>
+                        <key>name</key>
+                        <string>comment.block.carbon</string>
+                        <key>begin</key>
+                        <string>/\*</string>
+                        <key>end</key>
+                        <string>\*/</string>
+                    </dict>
+                    <!-- Unterminated multi-line comments -->
+                    <dict>
+                        <key>name</key>
+                        <string>invalid.illegal.unterminated-multi-line.carbon
+                        </string>
+                        <key>match</key>
+                        <string>\*/</string>
+                    </dict>
```
```suggestion
```

Sorry, I missed this before: there's no multi-line comment syntax at present.

jonmeow commented on utils/textmate/Samples/main.carbon:
```diff
@@ -0,0 +1,52 @@
+
+// More single line comments
+
+/* Multi line and closing comments */
+/* Spans
+   Multiple
+   Lines
+*/
+
```
```suggestion
```

Sorry, I'd missed this -- this isn't supported syntax.

jonmeow commented on utils/textmate/Samples/main.carbon:
```diff
@@ -0,0 +1,52 @@
+   Lines
+*/
+
+package Carbon /* between */ api;
```
```suggestion
package Carbon api;
```

jonmeow reviewed (commented):
Sorry, I'd missed the multi-line comments before. There are currently only single-line comments right now:
https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design#code-and-comments

We're trying that out for simplicity right now. I'm hesitant to add multi-line comment highlighting while we don't have them.

Other than that, I think this looks good now.

junlarsen committed:
fix: remove multilines

junlarsen commented:
> Sorry, I'd missed the multi-line comments before. There are currently only single-line comments right now: [`trunk`/docs/design#code-and-comments](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design?rgh-link-date=2022-07-28T17%3A41%3A20Z#code-and-comments)
> 
> We're trying that out for simplicity right now. I'm hesitant to add multi-line comment highlighting while we don't have them.
> 
> Other than that, I think this looks good now.

Oh that's interesting Bad assumption on my end, should have dove deeper into the design docs first. Fixed!

jonmeow reviewed (approved):
Thank you!
